### PR TITLE
CM-68: Develop core logic for handling custom filters - BE

### PR DESCRIPTION
**Pull request by sniedzielski**
TICKET: https://openimis.atlassian.net/browse/CM-68

PR delivers:
- Interface for CustomFilterWizard
- Registry point to obtain from particular module (on its apps.py file) information about implemented above interfaces while core module initialization
- Core Hub (storage) to build output how to build filters for particular object type 

The way how it works is below. This is the concept how it could work. Based on mocked fixed implementation of interfaces in social_protection module (real logic will be implemented within CM-69 ticket). If you have some ideas how to improve this solution at this stage please let me know in the review comments. 

- mocked implementation of interfaces for creation of custom filters
Screenshot from 2023-05-26 11-10-37
- mocked custom_filter.py file in policy module
Screenshot from 2023-05-31 09-41-45
Screenshot from 2023-05-31 21-02-34

- some operations on that service responsible for creating output based on given context from mocked implementation of services etc
Screenshot from 2023-05-31 21-07-17
Screenshot from 2023-05-31 21-11-20

NOTE: While working on CM-69 I will create graphQL query which will be necessary for frontend (CM-71) in order to retrieve such information how to build such filters per particular object type (for example how to build filters based on data provided from schema defined in BenefitPlan(CM-69))

2nd NOTE: We could think about Registry point as a generic solution. Similar actions happens inside registering calculation rules in calculation modules, registering appscheduler tasks, signals and grapgQL schema within assembly module. But this could be a proposal of improvement when we have the time and space for such improvements.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -352,6 +352,36 @@ CLASS_RULE_PARAM_VALIDATION = [
   - run_convert(instance, convert_to **argv) - execute the conversion for the instance with the first rule that provide the conversion (see get_convert_from_to). 'from_to' is deducted from instance. 
   - get_convert_from_to() - get the possible conversion, return [calc UUID, from, to]
 
+## CustomFilters
+This module provides the ability to define custom filters and their construction. It includes the following classes:
+* ```CustomFilterRegistryPoint```: This class is responsible for registering a specific module.
```
Comment (dborowiecki) [inline README.md:357]: `
```suggestion
* ```CustomFilterRegistryPoint```: This class is responsible for registering filters from a specific module. 
```

File: README.md [hunk 2]
```diff
@@ -352,6 +352,36 @@ CLASS_RULE_PARAM_VALIDATION = [
   - run_convert(instance, convert_to **argv) - execute the conversion for the instance with the first rule that provide the conversion (see get_convert_from_to). 'from_to' is deducted from instance. 
   - get_convert_from_to() - get the possible conversion, return [calc UUID, from, to]
 
+## CustomFilters
+This module provides the ability to define custom filters and their construction. It includes the following classes:
+* ```CustomFilterRegistryPoint```: This class is responsible for registering a specific module. 
+It should be called in the apps.py file of the respective module (see below) to enable filter construction for specific objects.
```
Comment (dborowiecki) [inline README.md:358]:
```suggestion
It should be called in the apps.py file of the respective module (see example below) to enable filter construction for specific objects.
```

File: README.md [hunk 3]
```diff
@@ -352,6 +352,36 @@ CLASS_RULE_PARAM_VALIDATION = [
...
+* ```CustomFilterWizardInterface```: This interface defines the implementation details of a custom filtering wizard. 
+Each module may have different implementations for delivering the desired output.
```
Comment (dborowiecki) [inline README.md:360]:
```suggestion
```
This one seems to be redundant

File: README.md [hunk 4]
```diff
@@ -352,6 +352,36 @@ CLASS_RULE_PARAM_VALIDATION = [
...
+* ```CustomFilterWizardStorage```: This class provides a service that connects to the registered filter hub, allowing 
+retrieval of the filter construction definition for specific objects.
+
+To integrate this mechanism into an openIMIS module, follow these steps:
+* Implement the CustomFilterWizardInterface from the core module (if no custom_filters.py file exists, create one in the module)
```
Comment (dborowiecki) [inline README.md:365]:
```suggestion
* Implement the CustomFilterWizardInterface from the core module.
```
Is this file still required?

File: README.md [hunk 5]
```diff
@@ -352,6 +352,36 @@ CLASS_RULE_PARAM_VALIDATION = [
...
+* ```CustomFilterWizardInterface```: This interface defines the implementation details of a custom filtering wizard. 
+Each module may have different implementations for delivering the desired output.
+* ```CustomFilterWizardStorage```: This class provides a service that connects to the registered filter hub, allowing 
+retrieval of the filter construction definition for specific objects.
+
+To integrate this mechanism into an openIMIS module, follow these steps:
+* Implement the CustomFilterWizardInterface from the core module (if no custom_filters.py file exists, create one in the module) 
+to output the filter construction definition for the chosen object.
+* In the apps.py file of the module, add the following import statement: ```from core.custom_filters import CustomFilterRegistryPoint```.
+* Within the module's configuration class, in the configuration loading section, include the following lines of code:
```
Comment (dborowiecki) [inline README.md:368]: I think that those 2 steps are not align in terms of "level of complexity". First one tells user excatly how to do the import and second makes assumption that it's obvious what is configuration loading section. You can say that it has to be invoked somewhere in ready()

File: core/custom_filters/custom_filter_registry_point.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+            This method registers the provided list of objects as custom filters in the hub,
+            for the specified module in the openIMIS application.
+
+            Args:
+                module_name (str): The name of the module installed in openIMIS.
+                custom_filter_class_list (list): A list of objects representing the custom filter implementations.
+
+            Returns:
+                None: This method does not return anything.
```
Comment (dborowiecki) [inline core/custom_filters/custom_filter_registry_point.py:50]: Correct me if I'm wrong but I think we're using reStructuredText format for dosctrings. If that's the case then please fix them

File: core/custom_filters/custom_filter_wizard_storage.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    __VALUE = 'value'
+
+    @classmethod
+    def build_output_how_to_build_filter(cls, module_name: str, object_type: str, **kwargs) -> List[namedtuple]:
+        """
+            Build the final outcome for creating a filter based on the information provided by
+            a registered custom filter wizard for the specified module name and object type.
+
+            The output is a list of named tuples (from the collections package). Each named tuple is built
+            in the following format: <Type>(field=<str>, filter=<str>, value=<str>)
```
Comment (dborowiecki) [inline core/custom_filters/custom_filter_wizard_storage.py:39]: I think we changed "value" to "type"

File: core/custom_filters/custom_filter_wizard_storage.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        implementing the wizard interface in the given object type.
+
+        Args:
+            wizard_filter_class (CustomFilterWizardInterface): An instance of the class responsible for
+                                                              implementing the CustomFilterWizardInterface
+                                                              in the given object type.
+            object_type (str): The string representation of the object class that takes part in filtering customization.
+
+        Returns:
+            bool: True if the object is matched, False otherwise.
```
Comment (dborowiecki) [inline core/custom_filters/custom_filter_wizard_storage.py:116]:
```suggestion
```

There's no need for documentation of private functions that do single operation. You could give a single line of explanation why given check is done, but this docstring is excessive.

Comment (dborowiecki) [reply core/custom_filters/custom_filter_wizard_storage.py:116]: It's the same for __create_instance_of_wizard_class

File: core/custom_filters/custom_filter_wizard_storage.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                List[namedtuple]: A list of named tuples representing the final outcome for creating a filter.
+        """
+        output_of_possible_filters = []
+        registered_filter_wizards = CustomFilterRegistryPoint.REGISTERED_CUSTOM_FILTER_WIZARDS
+        if module_name in registered_filter_wizards:
+            for registered_filter_wizard in registered_filter_wizards[module_name]:
+                if cls.__KEY_FOR_OBTAINING_CLASS in registered_filter_wizard:
+                    wizard_filter_class = cls.__create_instance_of_wizard_class(registered_filter_wizard)
+                    if cls.__check_object_type(wizard_filter_class, object_type):
+                        cls.__run_load_definition_object_in_wizard(wizard_filter_class, output_of_possible_filters, **kwargs)
```
Comment (dborowiecki) [inline core/custom_filters/custom_filter_wizard_storage.py:58]:
```suggestion
            for registered_filter_wizard in registered_filter_wizards[module_name]:
                if not cls.__KEY_FOR_OBTAINING_CLASS in registered_filter_wizard:
                    continue
                wizard_filter_class = cls.__create_instance_of_wizard_class(registered_filter_wizard)
                if not cls.__check_object_type(wizard_filter_class, object_type):
                    continue
                 cls.__run_load_definition_object_in_wizard(wizard_filter_class, output_of_possible_filters, **kwargs)
```
You can revert the ifs to avoid nesting

File: core/custom_filters/custom_filter_wizard_storage.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+        Returns:
+            None: This is a void method.
+        """
+        wizard_filter_tuple_type = namedtuple(
+            wizard_filter_class.get_type_of_object(),
+            [cls.__FIELD, cls.__FILTER, cls.__VALUE]
+        )
+        tuple_list_result = wizard_filter_class.load_definition(wizard_filter_tuple_type, **kwargs)
+        output_of_possible_filters.extend(tuple_list_result)
```
Comment (dborowiecki) [inline core/custom_filters/custom_filter_wizard_storage.py:87]: Instead of extending list in here you can just return `tuple_list_result` and add it to list in `build_output_how_to_build_filter`

File: core/custom_filters/custom_filter_wizard_storage.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+            __VALUE: Identifies the third value of the final output tuple for building filters.
+    """
+
+    __KEY_FOR_OBTAINING_CLASS = 'class_reference'
+    __FIELD = 'field'
+    __FILTER = 'filter'
+    __VALUE = 'value'
+
+    @classmethod
+    def build_output_how_to_build_filter(cls, module_name: str, object_type: str, **kwargs) -> List[namedtuple]:
```
Comment (dborowiecki) [inline core/custom_filters/custom_filter_wizard_storage.py:33]: I don't quite like this name, perhaps something like `build_custom_filters_definition` would be better.

**Review by dborowiecki — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (sniedzielski) [reply README.md:365]: No, this file is not required anymore according to what we discussed.

File: core/custom_filters/custom_filter_wizard_storage.py (hunk 7 above)
Comment (sniedzielski) [reply core/custom_filters/custom_filter_wizard_storage.py:39]: yes, I forgot to change here

Comment (sniedzielski) [reply core/custom_filters/custom_filter_wizard_storage.py:39]: Anyway It is changed in CM-69 PR

File: core/custom_filters/custom_filter_registry_point.py (hunk 6 above)
Comment (sniedzielski) [reply core/custom_filters/custom_filter_registry_point.py:50]: Yes, I'm going to fix this to use reStructuredText

**Review by dborowiecki — APPROVED** [review]
